cellarium/ml/models/cellarium_gpt.py

```diff
@@ -0,0 +1,774 @@
+        #   > A modified initialization which accounts for the accumulation on the residual path with model depth. Scale
+        #   > the weights of residual layers at initialization by a factor of 1/√N where N is the # of residual layers.
+        #   >   -- GPT-2 :: https://openai.com/blog/better-language-models/
+        #
+        # Reference (Megatron-LM): https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/model/gpt_model.py
+        for name, p in module.named_parameters():
+            if name == "dense2.weight" or name == "Wo.weight":
+                # Special Scaled Initialization --> There are 2 Layer Norms per Transformer Block
+                p.data.normal_(mean=0.0, std=(self.initializer_range / math.sqrt(2 * self.gpt_model.n_blocks)))
+
+    def tokenize(
+        self, x_ng: torch.Tensor, total_mrna_umis_n: torch.Tensor | None, context_size: int, shuffle: bool
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        n = x_ng.shape[0]
+        device = x_ng.device
+        ndx = torch.arange(n, device=device)
+
+        genes_context_size = context_size if total_mrna_umis_n is None else context_size - 1
+        if shuffle:
+            indices_ng = torch.argsort(torch.rand_like(x_ng))
+            indices_nc = indices_ng[:, :genes_context_size]
+        else:
+            indices_nc = torch.arange(genes_context_size, device=device).expand(n, -1)
+        values_nc = x_ng[ndx[:, None], indices_nc]
+        if total_mrna_umis_n is not None:
+            # concatenate total_mrna_umis to the random subset of genes
+            values_nc = torch.cat([total_mrna_umis_n[:, None], values_nc], dim=1)
+            ids_nc = torch.cat([torch.zeros((n, 1), dtype=torch.long, device=device), indices_nc + 1], dim=1)
+        else:
+            ids_nc = indices_nc + 1
+        return ids_nc, values_nc
+
+    def forward(
+        self, x_ng: torch.Tensor, var_names_g: np.ndarray, total_mrna_umis_n: torch.Tensor
+    ) -> dict[str, torch.Tensor]:
+        assert_columns_and_array_lengths_equal("x_ng", x_ng, "var_names_g", var_names_g)
+        assert_arrays_equal("var_names_g", var_names_g, "var_names_g", self.var_names_g)
+
+        n = x_ng.shape[0]
+        device = x_ng.device
+
+        # prefix includes the total_mrna_umis and a random subset of genes
+        # the length of the prefix is sampled uniformly from 1 to n_context - 1 (inclusive)
+        prefix_len_n = torch.randint(1, self.n_context, (n,), device=device)
+        labels_mask_nc = torch.arange(self.n_context, device=device)[None, :] >= prefix_len_n[:, None]
+
+        ids_nc, values_nc = self.tokenize(x_ng, total_mrna_umis_n, self.n_context, shuffle=True)
+        hidden_state_ncd = self.gpt_model(ids_nc, values_nc, labels_mask_nc, prefix_len_n, "block_diagonal")
+        mu_nc, theta_nc = self.nb_head(hidden_state_ncd)
+
+        sample_weights_nc = 1 / labels_mask_nc.sum(dim=1, keepdim=True).expand(-1, self.n_context)
```

Comment:
Is it true that with no masking for cell 0, i.e. `labels_mask_nc[0, :].sum(1) == 0`, then `sample_weights_nc[0, :] == inf` ?  Is this desired?

Reply:
I suppose there is always some masking during training.  I was thinking about prediction where maybe I would not mask anything, and I was seeing `inf` values (though I would not use `sample_weights` anyway).

Reply:
Oh I see you've enforced this anyway by saying the max range of the prefix size is `n_context - 1`, looks like you'll never have that 1 / 0 situation.

Reply:
That's right. `sample_weights_nc` is only needed during training to calculate the loss.